feat: use cbor4ii as underlying library

vmx opened the pull request:
cbor4ii is faster and nicer code than the old code base. It also has the
CBOR encoding/decoding separate from the Serde implementation, this means
that we only need our custom Serde implementation and can use the core
CBOR wrangling from cbor4ii.

BREAKING CHANGE: Errors and some API might have changed

It has less functionality, but the APIs that were broadly used should've
stayed the same. The errors are now split into Encode and Decode errors.

vmx committed:
feat: use cbor4ii as underlying library

cbor4ii is faster and nicer code than the old code base. It also has the
CBOR encoding/decoding separate from the Serde implementation, this means
that we only need our custom Serde implementation and can use the core
CBOR wrangling from cbor4ii.

BREAKING CHANGE: Errors and some API might have changed

It has less functionality, but the APIs that were broadly used should've
stayed the same. The errors are now split into Encode and Decode errors.

vmx commented:
I should've added that this PR is only about switching to cbor4ii, it should behave the same as the current code. Making the (de)serializing stricter is the next step and will be done in subsequent PRs.

dignifiedquire commented on Cargo.toml:
```diff
@@ -15,20 +15,16 @@ categories = ["encoding"]
-unsealed_read_write = []
-_do_not_use_its_unsafe_and_invalid_cbor = []
+std = ["cbor4ii/use_std", "cid/std", "serde/std", "serde_bytes/std"]
+_do_not_use_its_unsafe_and_invalid_cbor = ["std"]
```
I can haz comment for this feature please?

vmx replied:
Intentionally not. This should not be used. I don't want to encourage usage.

dignifiedquire replied:
I understand, but it would still be nice why/what it does

vmx replied:
I'd like to prevent people from finding it. It is only used by a single project and (as I understand it) they will not need it in the future, so it will be removed then. I can picture people using it if it's easier to find.

dignifiedquire commented on src/cbor4ii_nonpub.rs:
```diff
@@ -0,0 +1,70 @@
+
+// Copy from cbor4ii/util.rs.
+/// Executes the given function when the variables moves out of scope.
+pub(crate) struct ScopeGuard<'a, T>(pub &'a mut T, pub fn(&mut T));
```
rather than maintaining one, I would recommend using https://docs.rs/scopeguard

vmx replied:
Oh nice! I didn't know about that.

vmx replied:
@dignifiedquire can you help with that one? I'm running into lifetime issues when using `scopeguard`.

dignifiedquire replied:
looking

dignifiedquire replied:
https://github.com/vmx/serde_ipld_dagcbor/pull/14

vmx replied:
Thanks! Lol, all I was missing in my attempts was a single star.

dignifiedquire commented on src/de.rs:
```diff
@@ -117,965 +70,516 @@ where
-    accept_standard_enums: bool,
-    accept_legacy_enums: bool,
+/// A Serde `Deserialize`r of DAG-CBOR data.
+struct Deserializer<R> {
```
Debug?

dignifiedquire commented on src/de.rs:
```diff
@@ -117,965 +70,516 @@ where
-        scratch: &'b mut [u8],
-    ) -> Deserializer<SliceReadFixed<'a, 'b>> {
-        Deserializer::new(SliceReadFixed::new(bytes, scratch))
+    pub fn new(reader: R) -> Deserializer<R> {
```
why not keep this as `from_reader`?

vmx replied:
Symmetry to the serializer. But as this is a public API, there's no point of breaking it.

dignifiedquire commented on src/de.rs:
```diff
@@ -117,965 +70,516 @@ where
+            major::TAG => match byte {
+                0xc2 => de.deserialize_u128(visitor),
+                0xc3 => de.deserialize_i128(visitor),
+                0xd8 => {
```
can you use constants for these please?

dignifiedquire commented on src/de.rs:
```diff
@@ -117,965 +70,516 @@ where
+        #[cfg(feature = "_do_not_use_its_unsafe_and_invalid_cbor")]
+        match types::BadStr::<Cow<[u8]>>::decode(&mut self.reader)? {
+            types::BadStr(Cow::Borrowed(buf)) => {
+                visitor.visit_borrowed_str(unsafe { std::str::from_utf8_unchecked(buf) })
```
😭

dignifiedquire commented on src/de.rs:
```diff
@@ -117,965 +70,516 @@ where
-            }
-        })
+        let byte = pull_one(&mut self.reader)?;
+        // 0 length array
```
is this the only possibility, what about 0 length other things?

dignifiedquire replied:
also, why is an empty array encoded as unit oO

vmx replied:
That's `cbor4ii` in order to be able to distinguish between `Option::None` and unit. But we serialize it as `NULL`. Good catch, I've changed the code and also added a test for it.

dignifiedquire commented on src/de.rs:
```diff
@@ -117,965 +70,516 @@ where
+        V: Visitor<'de>,
     {
-        self.parse_value(visitor)
+        self.deserialize_map(visitor)
```
why does `struct` map to `map`?

vmx replied:
Because structs are serialized as CBOR maps.

dignifiedquire commented on src/de.rs:
```diff
@@ -1362,70 +640,63 @@ where
+        let byte = peek_one(&mut self.0.reader)?;
+        match dec::if_major(byte) {
+            major::BYTES => {
+                // CBOR encoded CIDs have a zero byte prefix we have to remove.
```
why?

vmx replied:
Why CBOR encodes CIDs with a zero byte prefix? This has historic reasons. CIDs are specified as being Multibase encoded. And the prefix `0x00` stands for "it's not encoded, use it as is". Though these days I don't see anyone considering CIDs being Multibase encoded, but rather speaks of "Multibase encoded CIDs" in case they are. This is the reason, why the CID implementation we are using (and any other CID implementation I know of) doesn't contain the `0x00` when returning a CID.

dignifiedquire replied:
😭

dignifiedquire commented on src/ser.rs:
```diff
@@ -401,115 +209,113 @@ where
         self,
-        name: &'static str,
-        variant_index: u32,
+        _name: &'static str,
```
why are these ignored now?

vmx replied:
`serde_cbor` had an option for a "packed" representation, where they weren't represented as maps. We never used that packed representation.

dignifiedquire commented on src/ser.rs:
```diff
@@ -401,115 +209,113 @@ where
-            Ok(self)
-        }
+    ) -> Result<Self::SerializeTupleVariant, Self::Error> {
+        enc::MapStartBounded(1).encode(&mut self.writer)?;
```
why a map with array?

vmx replied:
The data is a tuple => array, it has a name (the `variant`) => wrapped in a single item map. So

```rust
enum MyEnum {
    Data(u8, bool)
}
```

A `MyEnum::Data(7, true)`, becomes (I write it as JSON):

```json
{"Data": [7, true]}
```

vmx replied:
I forgot to mention that it's the same as the `self.enum_as_map == true` code path in the old code.

dignifiedquire replied:
I see, thanks, maybe adding a comment about this would be useful for future users

dignifiedquire commented on src/ser.rs:
```diff
@@ -401,115 +209,113 @@ where
-        self.serialize_unit_variant(name, variant_index, variant)?;
-        self.serialize_struct(name, len)
+    ) -> Result<Self::SerializeStructVariant, Self::Error> {
+        enc::MapStartBounded(1).encode(&mut self.writer)?;
```
why are you creating two maps?

vmx replied:
`MyStruct::Pointer { pos: 4, len: 2 }` becomes:

```json
{
  "Pointer": {
    "pos": 4,
    "len": 2
  }
}
```

dignifiedquire replied:
I see

dignifiedquire committed:
switch to use scopeguard crate

vmx committed:
fixup: apply changes from code review

dignifiedquire commented on src/de.rs:
```diff
@@ -165,12 +169,9 @@ impl<'de, 'a, R: dec::Read<'de>> serde::Deserializer<'de> for &'a mut Deserializ
-            // NOTE: that this does not support untagged enum.
-            // see https://github.com/serde-rs/serde/issues/1682
             major::TAG => match byte {
-                0xc2 => de.deserialize_u128(visitor),
```
are you removing support for 128 bit numbers ?

vmx replied:
Yes, they have a different tag => not valid DAG-CBOR.

dignifiedquire replied:
😭

vmx committed:
fixup: make Clippy happy

vmx committed:
fixup: even happier Clippy

vmx commented:
Also finally the FVM is green with this branch: https://github.com/filecoin-project/ref-fvm/pull/472

Stebalien commented on src/de.rs:
```diff
@@ -117,965 +73,517 @@ where
+            major::ARRAY => de.deserialize_seq(visitor),
+            major::MAP => de.deserialize_map(visitor),
+            major::TAG => match byte {
+                // Only TAG 42 (CID) is supported, which fits into 1 byte.
```
So, are we saying that it's _invalid_ if someone doesn't minimally encode the tag? Let's just accept non-minimally encoded tags as well.

Stebalien commented on src/de.rs:
```diff
@@ -117,965 +73,517 @@ where
-                Err(de.error(ErrorCode::TrailingData))
-            } else {
-                Ok(value)
+        if name == CID_SERDE_PRIVATE_IDENTIFIER {
```
ditto above. If we're going to accept all valid CBOR (modulo indefinite length lists, maybe), we should accept all valid CBOR (including non-minimally encoded tags).

Stebalien commented on src/de.rs:
```diff
@@ -1362,70 +644,63 @@ where
 /// The reason for not including the CBOR tag information is the [`Value`] implementation. That one
 /// starts to parse the bytes, before we could interfere. If the data only includes a CID, we are
 /// parsing over the tag to determine whether it is a CID or not and go from there.
+//struct CidDeserializer<'a, 'b, R: dec::Read<'b>>(&'a mut Deserializer<'b, R>);
```
remove?

Stebalien commented on src/de.rs:
```diff
@@ -1362,70 +644,63 @@ where
+                        if buf.len() <= 1 {
+                            Err(DecodeError::Msg("Invalid CID".into()))
+                        } else {
+                            buf.remove(0);
```
Just treating this as borrowed is likely cheaper than removing a byte and shifting everything over.

Stebalien reviewed (approved):
Some concerns over minimal versus non-minimal encoding, but otherwise LGTM (as far as I can tell, there's a lot of code).

Stebalien commented:
It's also a bit unfortunate that we need to replicate so much code from cbor4ii. Maybe we could convince them to expose more? Or maybe we could use their `Deserializer` under the covers, but override the tag handling?

vmx committed:
fixup: address code review

vmx commented:
@Stebalien please review again, I think I've addressed all issues. Please note that I switched to cbor4ii master as I needed an upstream change (which was merged). Once you're OK with this PR I'll ask the cbor4ii author for a release (I don't want to do it earlier, in case we need more changes, I don't want to stress their time too much).

Stebalien commented on src/de.rs:
```diff
@@ -117,12 +114,13 @@ impl<'de, R: dec::Read<'de>> Deserializer<R> {
             _ => Err(DecodeError::TypeMismatch {
                 name: "CBOR tag",
-                byte: tag_value,
+                byte: tag.0 as u8,
```
Adding a new enum variant for "UnsupportedTag" may be better. Otherwise, we're truncating the tag here and that could be very confusing (e.g., if the bottom byte is 42).

vmx replied:
I've just copied the errors from cbor4ii. There is an issue open about better errors (https://github.com/quininer/cbor4ii/issues/9). I agree it could be improved, but I'd rather leave this to future PRs.

Stebalien replied:
Works for me.

vmx committed:
fixup: use released version
